File: split/getavail.pas
function getavail: halfword;
var
  p: halfword;
begin
  p := avail;
  if p <> 0 then
    avail := mem[avail].hh.rh
  else if memend < memmax then
  begin
    memend := memend + 1;
    p := memend;
  end
  else
  begin
    himemmin := himemmin - 1;
    p := himemmin;
    if himemmin <= lomemmax then
    begin
      runaway;
      overflow(300, memmax + 1 - memmin);
    end;
  end;
  mem[p].hh.rh := 0;
  {dynused:=dynused+1;}
getavail := p;
end;

File: cpp/getavail.cpp
#include "getavail.h"
#include "runaway.h"
#include "overflow.h"

halfword getavail(void)
{
	halfword p = avail;
	if (p)
		avail = link(avail);
	else 
		if (memend < memmax)
			p = ++memend;
		else
		{
			p = --himemmin;
			if (himemmin <= lomemmax)
			{
				runaway();
				overflow("main memory size", memmax+1-memmin);
			}
		}
	link(p) = 0;
	return p;
}
